content/tutorials/build-apps/build-a-desktop-wallet-in-javascript.md

```diff
@@ -0,0 +1,1433 @@
+
+1. In the `library` folder, add a new file `5_helpers.js` with the following content:
+
+{{ include_code("_code-samples/build-a-wallet/desktop-js/library/5_helpers.js", language="js") }}
+
+2. Modify the import section at the top of `index.js` to look like this:
+
+```javascript
+const xrpl = require("xrpl")
+const { initialize, subscribe, saveSaltedSeed, loadSaltedSeed } = require('./library/5_helpers')
+
+const TESTNET_URL = "wss://s.altnet.rippletest.net:51233"
+
+const WALLET_DIR = 'Wallet'
+
+const createWindow = () => {
+```
+
+Note that we have reduced the imports to one line. The helper functions that we have written before now get used in our new helper class, which not only adds new functionality but encapsulates our subscriptions and initial requests in two helper functions. Those will be used as one-liners replacing a lot of lines that started to bloat our main logic file. 
+
+We also added a new constant containing the directory name where we are going to store our encrypted seed.
+
+3. In `index.js` replace the existing `main` function with the following one:
+
+```javascript
+const main = async () => {
+  const appWindow = createWindow()
+
+  // Create Wallet directory in case it does not exist yet
+  if (!fs.existsSync(WALLET_DIR)) {
+    fs.mkdirSync(path.join(__dirname, WALLET_DIR));
+  }
+
+  let seed = null;
+
+  ipcMain.on('seed-entered', async (event, providedSeed) => {
+    seed = providedSeed
+    appWindow.webContents.send('open-password-dialog')
+  })
+
+  ipcMain.on('password-entered', async (event, password) => {
+    if (!fs.existsSync(path.join(__dirname, WALLET_DIR , 'seed.txt'))) {
+      saveSaltedSeed('../' + WALLET_DIR, seed, password)
+    } else {
+      seed = loadSaltedSeed('../' + WALLET_DIR, password)
+    }
+
+    const wallet = xrpl.Wallet.fromSeed(seed)
+
+    const client = new xrpl.Client(TESTNET_URL)
+
+    await client.connect()
+
+    await subscribe(client, wallet, appWindow)
+
+    await initialize(client, wallet, appWindow)
+  })
+
+  // We have to wait for the application frontend to be ready, otherwise
+  // we might run into a race condition and the open-dialog events
+  // get triggered before the callbacks are attached
+  appWindow.once('ready-to-show', () => {
+    // If there is no seed present yet, ask for it, otherwise query for the password
+    // for the seed that has been saved
+    if (!fs.existsSync(path.join(__dirname, WALLET_DIR, 'seed.txt'))) {
+      appWindow.webContents.send('open-seed-dialog')
+    } else {
+      appWindow.webContents.send('open-password-dialog')
+    }
+  })
+}
+```
+
+The subscription to the `address-entered` frontend is gone, as we are now going to use a full-fledged wallet, for which we are going to prompt the user or a seed and a password to encrypt it with. If there already is a seed, the user will only be asked for his password which is in turn used to decrypt the seed.
```

Comment:
```suggestion
Since we are now making this a full-fledged wallet, instead of asking the user for an address we will now be prompting the user for a seed and password to encrypt the seed. If there is already a seed, the user will only be asked for their password.```